File: solutions/46-permutations/permutations.py
class Solution:
    def solve(self, nums: list[int], l: int, r: int, result: list[list[int]]) -> None:
        # Base case: when left index equals right index
        if l == r:
            result.append(nums[:])  # Add copy of current permutation
            return
        
        # Try all possible elements at position 'l'
        for i in range(l, r + 1):
            # 1. Swap current element with element at position 'l'
            nums[l], nums[i] = nums[i], nums[l]
            
            # 2. Recursively generate permutations for remaining elements
            self.solve(nums, l + 1, r, result)
            
            # 3. Backtrack: restore array to original state by swapping back
            nums[l], nums[i] = nums[i], nums[l]
    
    def permute(self, nums: list[int]) -> list[list[int]]:
        result = []  # Stores all permutations
        self.solve(nums, 0, len(nums) - 1, result)  # Start with full array range
        return result

File: solutions/46-permutations/permutations.cpp
class Solution {
public:
    // Helper function that generates all permutations using backtracking
    // nums: input array to permute
    // l: left index of current subarray being processed
    // r: right index of current subarray being processed
    // result: stores all generated permutations
    void solve(vector<int>& nums, int l, int r, vector<vector<int>>& result) {
        // Base case: when left index equals right index
        // meaning we've processed all positions in current permutation
        if (l == r) {
            result.push_back(nums);  // Add current permutation to result
            return;
        }
        
        // Try all possible elements at position 'l'
        for (int i = l; i <= r; i++) {
            // 1. Swap current element with element at position 'l'
            swap(nums[l], nums[i]);
            
            // 2. Recursively generate permutations for remaining elements
            solve(nums, l + 1, r, result);
            
            // 3. Backtrack: restore array to original state by swapping back
            swap(nums[l], nums[i]);
        }
    }

    // Main function that initializes the permutation generation
    // nums: input array to find permutations for
    // returns: vector containing all possible permutations
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;  // Stores all permutations
        solve(nums, 0, nums.size() - 1, result);  // Start with full array range
        return result;
    }
};